packages/cli/src/services/public-api-key.service.ts

```diff
@@ -134,17 +142,67 @@ export class PublicApiKeyService {
 		};
 	}
 
-	private generateApiKey = (user: User, expiresAt: UnixTimestamp) => {
+	private generateApiKey(user: User, expiresAt: UnixTimestamp) {
 		const nowInSeconds = Math.floor(Date.now() / 1000);
 
 		return this.jwtService.sign(
 			{ sub: user.id, iss: API_KEY_ISSUER, aud: API_KEY_AUDIENCE },
 			{ ...(expiresAt && { expiresIn: expiresAt - nowInSeconds }) },
 		);
-	};
+	}
 
 	private getApiKeyExpiration = (apiKey: string) => {
 		const decoded = this.jwtService.decode(apiKey);
 		return decoded?.exp ?? null;
 	};
+
+	apiKeyHasValidScopesForRole(role: GlobalRole, apiKeyScopes: ApiKeyScope[]) {
+		const scopesForRole = getApiKeyScopesForRole(role);
+		return apiKeyScopes.every((scope) => scopesForRole.includes(scope));
+	}
+
+	async apiKeyHasValidScopes(apiKey: string, endpointScope: ApiKeyScope) {
+		const apiKeyData = await this.apiKeyRepository.findOne({
+			where: { apiKey },
+			select: ['scopes'],
+		});
+		if (!apiKeyData) return false;
+
+		return apiKeyData.scopes.includes(endpointScope);
+	}
+
+	getApiKeyScopeMiddleware(endpointScope: ApiKeyScope) {
+		return async (req: Request, res: Response, next: NextFunction) => {
+			const apiKey = req.headers['x-n8n-api-key'] as string;
+
+			const valid = await this.apiKeyHasValidScopes(apiKey, endpointScope);
```

Comment:
Did we talk about caching the api keys? I think we talked about doing this if it becomes an issue, right?

Reply:
As far as I remember we did not, but if this becomes an issue we can always add some caching.